app/scripts/directives/bindService.js

```diff
@@ -152,28 +157,37 @@
       onShow: showResults
     };
 
-    var updateInstance = function() {
-      if (!ctrl.serviceClasses || !ctrl.servicePlans) {
+    var updateServiceInstance = function() {
+      if (!ctrl.serviceToBind) {
         return;
       }
 
-      var instance = ctrl.target.kind === 'ServiceInstance' ? ctrl.target : ctrl.serviceToBind;
-      if (!instance) {
-        return;
+      if ((ctrl.target.kind !== 'ServiceInstance') && !ctrl.serviceClasses && !serviceClassesPromise) {
+        serviceClassesPromise = DataService.list(serviceClassesVersion, {}).then(function(serviceClasses) {
+          ctrl.serviceClasses = serviceClasses.by('metadata.name');
+          serviceClassesPromise = null;
+          sortServiceInstances();
+        });
       }
 
-      var serviceClassName = ServiceInstancesService.getServiceClassNameForInstance(instance);
-      ctrl.serviceClass = ctrl.serviceClasses[serviceClassName];
-      var servicePlanName = ServiceInstancesService.getServicePlanNameForInstance(instance);
-      ctrl.plan = ctrl.servicePlans[servicePlanName];
-      ctrl.parameterSchema = _.get(ctrl.plan, 'spec.serviceBindingCreateParameterSchema');
-      ctrl.parameterFormDefinition = _.get(ctrl.plan, 'spec.externalMetadata.schemas.service_binding.create.openshift_form_definition');
-      bindParametersStep.hidden = !_.has(ctrl.parameterSchema, 'properties');
-      ctrl.nextTitle = bindParametersStep.hidden ? 'Bind' : 'Next >';
-      ctrl.hideBack = bindParametersStep.hidden;
+      ServiceInstancesService.fetchServiceClassForInstance(ctrl.serviceToBind).then(function(serviceClass) {
+        ctrl.serviceClass = serviceClass;
+        Catalog.getServicePlansForServiceClass(ctrl.serviceClass).then(function (plans) {
```

Comment:
If this isn't returning cached data, it's probably better to just request the plan by name.

```js
var servicePlansVersion = APIService.getPreferredVersion('clusterserviceplans');
var servicePlanName = ServiceInstancesService.getServicePlanNameForInstance(ctrl.serviceToBind);
DataService.get(servicePlansVersion, servicePlanName, {}).then(function(plan) {
  ctrl.plan = plan;
  // ...
});
```

Reply:
Although I guess we have to worry about a race if this can get called more than once before the first request completes :(

Reply:
The list of plans is need only if/when we support podPresets and allow the binding to occur from an application and then I believe we need all plans so this list would be incorrect anyways. I will move the list of plans retrieval to where it is needed.

This would run into a race condition either way right? This, again, would only happen when binding from an application and if the user were to switch service instances quickly.

Reply:
I'm starting to think we want different components for the different bind dialogs (with any common logic for creating the binding in a service), but that's a bigger change :/

Reply:
yeah, I think I agree with you there. It's gotten more complex and hardly seems worth it at this point.